File: elixir/hamming/hamming.exs
defmodule DNA do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> DNA.hamming_distance('AAGTCATA', 'TAGCGATC')
  4
  """
  @spec hamming_distance([char], [char]) :: non_neg_integer
  def hamming_distance(strand, strand), do: 0
  def hamming_distance(strand1, strand2) when length(strand1) != length(strand2), do: nil
  def hamming_distance(strand1, strand2) do
    Enum.zip(strand1, strand2)
    |> Enum.map(&compare_nucleotides/1)
    |> Enum.sum
  end

  defp compare_nucleotides({x, x}), do: 0
  defp compare_nucleotides({x, y}), do: 1
end

File: go/hamming/hamming.go
package hamming

import "fmt"

const testVersion = 5

// Calculates the Hamming difference between two DNA strands.
func Distance(strand1, strand2 string) (int, error) {
	if len(strand1) != len(strand2) {
		return 0, fmt.Errorf("strands must be equal lengths")
	}

	hammingDistance := 0

	for i, nucleotide := range strand1 {
		if nucleotide != rune(strand2[i]) {
			hammingDistance += 1
		}
	}

	return hammingDistance, nil
}
